plugins/Validation/src/Validation.ts

```diff
@@ -1,223 +1,294 @@
+            return buildProxy(target, instance, [...path, nestedProp], state, downgraded, condition);
+        },
+    });
 }
 
-// tslint:disable-next-line: function-name
 export function Validation(): Plugin;
-export function Validation<S>($this: State<S>): ValidationExtensions<S>;
-export function Validation<S>($this?: State<S>): Plugin | ValidationExtensions<S> {
-    if ($this) {
-        let state = $this;
-
-        const [plugin] = state.attach(PluginID);
-        if (plugin instanceof Error) {
-            throw plugin
-        }
-        const instance = plugin as ValidationPluginInstance<S>;
-
-        const inst = instance;
+export function Validation(input: State<string>): SingleValidator<string>;
+export function Validation(input: State<number>): SingleValidator<number>;
+export function Validation<T>(input: State<T[]>): ArrayValidator<T[], T>;
+export function Validation<T>(input: State<T>): ObjectValidator<T, T>;
+export function Validation<T>(input?: State<T>): Plugin | ReturnType<T, T> {
+    if (input === undefined) {
         return {
-            validate: (r, m, s) => {
-                inst.addRule(state.path, {
-                    rule: r,
-                    message: m,
-                    severity: s || 'error'
-                })
-            },
-            validShallow(): boolean {
-                return inst.getErrors(state, 1, undefined, true).length === 0
-            },
-            valid(): boolean {
-                return inst.getErrors(state, Number.MAX_SAFE_INTEGER, undefined, true).length === 0
-            },
-            invalidShallow(): boolean {
-                return inst.getErrors(state, 1, undefined, true).length !== 0
+            id: ValidationId,
+            init: () => {
+                return new ValidatorInstance() as PluginCallbacks;
             },
-            invalid(): boolean {
-                return inst.getErrors(state, Number.MAX_SAFE_INTEGER, undefined, true).length !== 0
-            },
-            errors: (filter, depth, first) => {
-                return inst.getErrors(state, depth === undefined ? Number.MAX_SAFE_INTEGER : depth, filter, first);
-            },
-            firstError: (filter, depth) => {
-                const r = inst.getErrors(state, depth === undefined ? Number.MAX_SAFE_INTEGER : depth, filter, true);
-                if (r.length === 0) {
-                    return {};
-                }
-                return r[0];
-            },
-        }
+        };
     }
-    return {
-        id: PluginID,
-        init: () => new ValidationPluginInstance() as {}
+
+    const [instance] = input.attach(ValidationId);
+
+    if (instance instanceof Error) {
+        throw new Error(`Forgot to run state.attach(Validation())`);
+    }
+
+    if (!(instance instanceof ValidatorInstance)) {
+        throw new Error('Expected plugin to be of ValidatorInstance');
     }
+
+    const downgraded: State<T> = useState(input);
```

Comment:
Hi, thanks for the code. I will have a look at it later. This comment is only after a quick scan. UseState function can be used only within another hook functions which is named useXxxx. So, using it here is not ok. I am not sure why you need downgraded state. You can get advanced access to the state object from the returned argument from attach method (see the second element in the array of type: https://hookstate.js.org/docs/typedoc-hookstate-core#interfacespluginstatecontrolmd). Using the controls you can read the state without leaving marks of it being used (which is dangerous, because you may end up with missed rerenders).

Reply:
ah, i missed that.. yeah, i needed a way to access the state in a way that gave me raw recursive object access so that validators don't have to call .get() on everything and can run as a normal POJO. i call state alongside the downgraded access, so missing state change wasn't an issue.